src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp

```diff
@@ -371,6 +375,16 @@ void ShenandoahControlThread::request_gc(GCCause::Cause cause) {
 }
 
 void ShenandoahControlThread::handle_requested_gc(GCCause::Cause cause) {
+  // For normal requested GCs (System.gc) we want to block the caller. However,
+  // for whitebox requested GC, we want to initiate the GC and return immediately.
+  // The whitebox caller thread will arrange for itself to wait until the GC notifies
+  // it that has reached the requested breakpoint (phase in the GC).
+  if (cause == GCCause::_wb_breakpoint) {
+    _requested_gc_cause = cause;
+    _gc_requested.set();
+    return;
+  }
```

Comment:
Does this relate to GenShen, or is it a general issue?

Reply:
Yes, I believe it _could_ also affect other modes. I'm not sure why we've never seen the failure fixed by this change upstream. I will prepare a separate upstream PR for this.

Reply:
https://bugs.openjdk.org/browse/JDK-8340400